### PR TITLE
fix(appset): handle templating of raw JSON fields (#12947)

### LINKED ISSUE #12947: Parameters of other child generator not being substituted in List generator when using Matrix generator
Normally when using the Matrix generator, you are able to use parameters from the first child generator in the second child generator, as documented here: [https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-Matrix/#using-parameters-from-one-c[…]-in-another-child-generator](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-Matrix/#using-parameters-from-one-child-generator-in-another-child-generator)

However, when using a List generator as the second generator in a Matrix generator, it is not possible to use parameters of the first generator. Instead, the `{{ parameter }}` is not substituted, but taken as the literal value.

**To Reproduce**
You can reproduce this with this example application set:
```
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: testset
  namespace: argocd
spec:
  generators:
    - matrix:
        generators:
          - git:
              directories:
                - path: applications/*
              repoURL: 'https://github.com/example-org/applications.git'
              revision: HEAD
          - list:
              elements:
                - testKey: 'val-{{ path.basename }}'
  template:
    metadata:
      name: app-{{ path.basename }}
    spec:
      destination:
        name: dev
        namespace: app-{{ path.basename }}
      project: default
      source:
        helm:
          releaseName: release-{{ path.basename }}
          values: |
            testKey: {{.testKey}}
        path: test
        repoURL: 'https://github.com/example-org/manifests.git'
        targetRevision: refs/heads/release/1.2.x
      syncPolicy:
        automated:
          prune: true
```
When a Git repository contains:
```
applications/a
applications/b
```
this yields:
```
testKey: val-{{ path.basename }}
```
for both applications.

**Expected behavior**

A `{{ parameter }}` is substituted with the value from the first child generator, yielding this for the two applications in the above example:
```
testKey: val-a
testKey: val-b
```

**Version**

```shell
argocd: v2.6.6+6d4de2e.dirty
  BuildDate: 2023-03-17T00:05:38Z
  GitCommit: 6d4de2ec5d49fa2c6823f2b7d101607a839be3fa
  GitTreeState: dirty
  GoVersion: go1.20.2
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.6.6+6d4de2e
```

**Slack discussion**
https://cloud-native.slack.com/archives/C01TSERG0KZ/p1679386558427959

### PR DESCRIPTION
**Pull request by crenshaw-dev**
Fixes #12947

### CODE REVIEW

**Comment by ishitasequeira** [comment]
Hi @crenshaw-dev , seeing the below error while running `make start`. Maybe we need to rebase the PR.

```
Step 30/34 : COPY --from=node /opt/yarn-v1.22.4 /opt/yarn-v1.22.4
COPY failed: stat opt/yarn-v1.22.4: file does not exist
make: *** [test-tools-image] Error 1
```

**Comment by crenshaw-dev** [comment]
Rebased, it should work now!

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Comment by sarsatis** [comment]
Hi crenshaw my argocd is running on below version
{
    "Version": "v2.6.7+5bcd846",
    "BuildDate": "2023-03-23T14:57:27Z",
    "GitCommit": "5bcd846fa16e4b19d8f477de7da50ec0aef320e5",
    "GitTreeState": "clean",
    "GoVersion": "go1.18.10",
    "Compiler": "gc",
    "Platform": "linux/amd64",
    "KustomizeVersion": "v4.5.7 2022-08-02T16:35:54Z",
    "HelmVersion": "v3.10.3+g835b733",
    "KubectlVersion": "v0.24.2",
    "JsonnetVersion": "v0.19.1"
}

I am not able to pass parametes passed by parent getting the below error
status:
  conditions:
    - lastTransitionTime: '2023-04-16T09:27:30Z'
      message: >-
        Application.argoproj.io "{{ path.basenameNormalized }}-dev" is invalid:
        metadata.name: Invalid value: "{{ path.basenameNormalized }}-dev": a
        lowercase RFC 1123 subdomain must consist of lower case alphanumeric
        characters, '-' or '.', and must start and end with an alphanumeric
        character (e.g. 'example.com', regex used for validation is
        '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
      reason: UpdateApplicationError
      status: 'True'
      type: ErrorOccurred
    - lastTransitionTime: '2023-04-16T09:16:22Z'
      message: Successfully generated parameters for all Applications
      reason: ParametersGenerated
      status: 'True'
      type: ParametersGenerated
    - lastTransitionTime: '2023-04-16T09:27:30Z'
      message: >-
        Application.argoproj.io "{{ path.basenameNormalized }}-dev" is invalid:
        metadata.name: Invalid value: "{{ path.basenameNormalized }}-dev": a
        lowercase RFC 1123 subdomain must consist of lower case alphanumeric
        characters, '-' or '.', and must start and end with an alphanumeric
        character (e.g. 'example.com', regex used for validation is
        '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
      reason: UpdateApplicationError
      status: 'False'
      type: ResourcesUpToDate

**Comment by sangdammad** [comment]
Seems like this PR has no issues being merged. When will this be approved and merged? It will really help our use case! Thanks

File: applicationset/utils/utils.go [hunk 1]
```diff
@@ -96,6 +96,25 @@ func (r *Render) deeplyReplace(copy, original reflect.Value, replaceMap map[stri
...
+					return fmt.Errorf("failed to deeply replace JSON field contents: %w", err)
+				}
+				jsonCopyInterface := jsonCopy.Interface().(*interface{})
+				data, err := json.Marshal(jsonCopyInterface)
+				if err != nil {
+					return fmt.Errorf("failed to marshal templated JSON field: %w", err)
+				}
+				copy.Field(i).Set(reflect.ValueOf(data))
 			} else if err := r.deeplyReplace(copy.Field(i), original.Field(i), replaceMap, useGoTemplate); err != nil {
 				return err
```
Comment (jannfis) [inline applicationset/utils/utils.go:119]: Just a small nit here: In the newly introduced code block, we return `fmt.Errorf("failed to deeply replace JSON field contents: %w", err)` when `r.deeplyReplace()` fails, I like that. Maybe we could streamline that error message to this  place as well.

**Review by jannfis — APPROVED** [review]
LGTM, just one tiny nit (on unrelated code :smile: )

**Comment by sangdammad** [comment]
thanks @jannfis for approving. @crenshaw-dev can this be merged please?

**Comment by crenshaw-dev** [comment]
Sorry @sangdammad this completely slipped my mind. Will get it cherry-picked and released.

**Pull request merged**